### PR TITLE
Enforce Config.DATADIR and Dataset._data_dir to be absolute

### PR DESCRIPTION
**Pull request by xuhdev**

### CODE REVIEW

File: pydax/_config.py [hunk 1]
```diff
@@ -29,6 +30,13 @@ class Config:
     # DATADIR is the default dir where datasets files are downloaded/loaded to/from.
     DATADIR: pathlib.Path = pathlib.Path.home() / '.pydax' / 'data'
 
+    def __post_init_post_parse__(self) -> None:
+        "This is called by :meth:`.__init__()` after data type validation."
+        # DATADIR should be absolute.
+        # We use object.__setattr__ because we set frozen=True, same as what dataclasses does:
+        # https://docs.python.org/3/library/dataclasses.html#frozen-instances
+        object.__setattr__(self, 'DATADIR', pathlib.Path(os.path.abspath(self.DATADIR)))
```
Comment (edwardleardi) [inline pydax/_config.py:38]: So `__init__` first sets `DATADIR: pathlib.Path = pathlib.Path.home() / '.pydax' / 'data'` and validates `pathlib.Path` type, then runs `__post_init_post_parse__`?

File: pydax/dataset.py [hunk 2]
```diff
@@ -71,7 +71,7 @@ def __init__(self, schema: SchemaDict,
         """
 
         self._schema: SchemaDict = schema
-        self._data_dir: pathlib.Path = pathlib.Path(data_dir)
+        self._data_dir: pathlib.Path = pathlib.Path(os.path.abspath(data_dir))
```
Comment (edwardleardi) [inline pydax/dataset.py:74]: Should we document better in the docstring an explanation of how to properly set the `data_dir`? The case I'm thinking of is where the user uses an incorrect absolute path, e.g. 

```python
import pydax
abs_path = 'edward@ibm.com/.pydax/data'  # when absolute path is really '/Users/edward@ibm.com/.pydax/data'
pydax.init(DATADIR=abs_path)
pydax.get_config()
# Config(DATADIR=PosixPath('/Users/edward@ibm.com/Desktop/IBM-Projects/DAX/DAX-API-Project/dax-api/edward@ibm.com/.pydax/data'))
```

File: tests/conftest.py [hunk 3]
```diff
@@ -15,13 +15,39 @@
...
+        symlink_dir = tmp_path / str(uuid.uuid4())
+        # A collision is extremely unlikely, but for the sake of completeness, this check never hurts.
+        if not symlink_dir.exists():
+            break
+
+    symlink_dir.symlink_to(tmp_sub_dir, target_is_directory=True)
+
+    yield symlink_dir
+
+    symlink_dir.unlink()
```
Comment (edwardleardi) [inline tests/conftest.py:48]: Does this line get run after a test using the `tmp_symlink_dir` fixture completes? Is it necessary?

File: tests/test_config.py [hunk 4]
```diff
@@ -43,6 +44,21 @@ def test_custom_data_dir(tmp_path, wikitext103_schema):
...
+def test_custom_relative_data_dir(tmp_path):
+    "Test using a custom relative data directory."
+
+    init(DATADIR=os.path.relpath(tmp_path))
+    assert get_config().DATADIR == tmp_path
+    assert get_config().DATADIR.is_absolute()
+
+
+def test_custom_symlink_data_dir(tmp_symlink_dir):
+    "Test using a custom symlink data directory. It shouldn't be resolved."
```
Comment (edwardleardi) [inline tests/test_config.py:56]: What do you mean by it shouldn't be resolved?

**Review by edwardleardi — APPROVED** [review]

File: pydax/_config.py (hunk 1 above)
Comment (xuhdev) [reply pydax/_config.py:38]: In the default case, yes. If a user specifies a relative path, then `DATADIR` got assigned first with that relative dir, then `__post_init_post_parse__` corrects it.

File: pydax/dataset.py (hunk 2 above)
Comment (xuhdev) [reply pydax/dataset.py:74]: I think absolute path should be a pretty well known concept among developers. Do you have some specific confusing point in mind?

File: tests/conftest.py (hunk 3 above)
Comment (xuhdev) [reply tests/conftest.py:48]: Yes, it runs after the fixture is done (this is after `yield`). Just clean up after being used.

File: tests/test_config.py (hunk 4 above)
Comment (xuhdev) [reply tests/test_config.py]: The symlink shouldn't be resolved. I've updated the docstring.

Comment (edwardleardi) [reply tests/test_config.py]: I meant more what does it mean for a symlink to resolve haha, but it's ok I looked it up. So we're testing that the symlink doesn't get replaced by the absolute path of what the symlink linked to?

File: tests/conftest.py (hunk 3 above)
Comment (edwardleardi) [reply tests/conftest.py:48]: Does `temp_dir` not handle that when it gets autodeleted at the end of a test?

File: pydax/dataset.py (hunk 2 above)
Comment (edwardleardi) [reply pydax/dataset.py:74]: Ok, its fine then. Do we want to mention that `data_dir` can be either relative or absolute or is that implied?

File: pydax/_config.py (hunk 1 above)
Comment (edwardleardi) [reply pydax/_config.py:38]: makes sense thanks!

File: tests/test_config.py (hunk 4 above)
Comment (xuhdev) [reply tests/test_config.py]: Yes, that's right

File: tests/conftest.py (hunk 3 above)
Comment (xuhdev) [reply tests/conftest.py:48]: Do you mean `tmp_path`? Yeah that should be generally true, but the pytest documentation about it is so vague that I'm not sure whether there could be some scenarios that `tmp_path` might not be deleted.

File: pydax/dataset.py (hunk 2 above)
Comment (xuhdev) [reply pydax/dataset.py:76]: Good point, I added some document

File: tests/test_config.py (hunk 4 above)
Comment (edwardleardi) [reply tests/test_config.py]: got it thanks

File: tests/conftest.py (hunk 3 above)
Comment (edwardleardi) [reply tests/conftest.py:48]: yeah meant `tmp_path`, got it can't hurt then

File: pydax/dataset.py (hunk 2 above)
Comment (edwardleardi) [reply pydax/dataset.py:76]: thanks

**Review by edwardleardi — APPROVED** [review]
LGTM thanks!

File: tests/conftest.py (hunk 3 above)
Comment (xuhdev) [reply tests/conftest.py:48]: FYI seems like `tmp_path` isn't cleaned up automatically: https://github.com/pytest-dev/pytest/issues/8036#issuecomment-727197675

We might have to adjust tests that download giant datasets.

**Pull request merged**